src/main/host/descriptor/descriptor_table.rs

```diff
@@ -0,0 +1,204 @@
+use super::CompatDescriptor;
+use crate::cshadow;
+use crate::utility::notnull::*;
+use log::*;
+use std::convert::TryInto;
+
+/// Table of (file) descriptors. Typically owned by a Process.
+pub struct DescriptorTable {
+    descriptors: std::collections::HashMap<u32, CompatDescriptor>,
+    available_indices: Vec<u32>,
+    index_counter: u32,
+}
```

Comment:
If we want to keep the available indices sorted so that we can always take the smallest, a BTreeSet might work well here.